pkg/controller/cassandrarestore/reconcile.go

```diff
@@ -0,0 +1,308 @@
+		return errors.WrapIfWithDetails(err, "could not update status for restore", "restore", restore)
+	}
+
+	return nil
+}
+
+func (r *ReconcileCassandraRestore) checkRestoreOperationState(restore *v1alpha1.CassandraRestore,
+	cc *v1alpha1.CassandraCluster, backup *v1alpha1.CassandraBackup, reqLogger *logrus.Entry) error {
+
+	pods, err := r.listPods(restore.Namespace, k8s.LabelsForCassandraDC(cc, backup.Spec.Datacenter))
+	if err != nil {
+		return errorfactory.New(errorfactory.ResourceNotReady{}, err, "no pods founds for this dc")
+	}
+
+	restoreId := restore.Status.ID
+	if restoreId == "" {
+		return errors.New("no Restore operation id provided to be checked")
+	}
+
+	// Check Restore operation status
+	sr, err := backrest.NewClient(r.client, cc, k8s.PodByName(pods, restore.Status.CoordinatorMember))
+	if err != nil {
+		reqLogger.Info("cassandra backup sidecar communication error checking running Operation", "OperationId",
+			restoreId)
+		return errorfactory.New(errorfactory.CassandraBackupSidecarNotReady{}, err,
+		"cassandra backup sidecar communication error")
+	}
+	restoreStatus, err := sr.RestoreStatusByID(restoreId)
+	if err != nil {
+		reqLogger.Info("cassandra backup sidecar communication error checking running Operation",
+			"OperationId", restoreId)
+		return errorfactory.New(errorfactory.CassandraBackupSidecarNotReady{}, err,
+		"cassandra backup sidecar communication error")
+	}
+
+	if err := UpdateRestoreStatus(r.client, restore, *restoreStatus, reqLogger); err != nil {
+		return errors.WrapIfWithDetails(err, "could not update status for restore",
+			"restore", restore)
+	}
+
+	// Restore operation failed or canceled,
+	// TODO : reschedule it by marking restore Condition.State = RestoreRequired ?
+	if restore.Status.Condition.Type.IsInError() {
+		return errorfactory.New(errorfactory.CassandraBackupOperationFailure{}, err,
+		"cassandra backup sidecar Operation failed", fmt.Sprintf("restore operation id : %s", restoreId))
+	}
+
+	// Restore operation completed successfully
+	if restore.Status.Condition.Type.IsCompleted() {
+		return nil
+	}
+
+	// TODO : Implement timeout ?
+
+	// restore operation still in progress
+	reqLogger.Info("Cassandra backup sidecar operation is still running", "restoreId", restoreId)
+	return errorfactory.New(errorfactory.CassandraBackupOperationRunning{},
+		errors.New("cassandra backup sidecar restore operation still running"),
+			fmt.Sprintf("restore operation id : %s", restoreId))
+}
+
+func (r *ReconcileCassandraRestore) updateAndFetchLatest(ctx context.Context,
+	restore *v1alpha1.CassandraRestore) (*v1alpha1.CassandraRestore, error) {
+	typeMeta := restore.TypeMeta
+	err := r.client.Update(ctx, restore)
+	if err != nil {
+		return nil, err
+	}
+	restore.TypeMeta = typeMeta
+	return restore, nil
+}
+
+func (r *ReconcileCassandraRestore) listPods(namespace string, selector map[string]string) (*v1.PodList, error) {
```

Comment:
this code is never re-used somewhere else ?

Reply:
there are identical functions but on different reconcile object. Maybe we should refactor it under as they all use the client of the reconcile object ?